### ISSUE TITLE
Memory allocation with CPU-only

### ISSUE THREAD

**Issue by jeffgortmaker** [google/jax#2023]
In #788 there's some talk about how to configure XLA memory usage with GPUs -- in particular, with the environment variables described in [this page](https://jax.readthedocs.io/en/latest/gpu_memory_allocation.html).

Are there similar configuration options when using only CPUs?

I tried setting `XLA_PYTHON_CLIENT_ALLOCATOR=platform` to see if this option carries over to CPU-only mode, but it doesn't get rid of the memory allocation request errors that I'm getting, such as:

```
Execution of replica 0 failed: Resource exhausted: Failed to allocate request for 321.75MiB (337385120B) on device ordinal 0
```
Only thing that seems to help so far is increasing the amount of memory available to my job (far beyond what it ends up using), which isn't ideal.

**Comment by skye** [comment]
I don't believe XLA slab allocates on CPU, meaning it should only be using what's needed. I could be wrong about this though. What are you seeing that indicates it's using more than necessary?

**Comment by jeffgortmaker** [comment]
I see, that's strange. I'm comparing the output of a [memory-profiler](https://pypi.org/project/memory-profiler/) call to how much memory I allocate to the job, and I assumed (wrongly I guess) that the above error message was from XLA trying to slab allocate.

But if XLA doesn't slab allocate on CPU, then the issue is more likely to be with how I'm profiling memory usage. Feel free to close unless you have any other thoughts - thanks!

**Comment by skye** [comment]
I've noticed that memory-profiler underreports the process's total memory usage, I suspect it's somehow not counting the memory used by our C++ XLA bindings (that's a total guess though). I'm gonna close this issue, but please keep commenting if you find out anything more.

**Comment by jeffgortmaker** [comment]
I realized I never followed on up this. In case anyone else has a similar issue with memory-profiler, I ended up finding that the builtin [resource](https://docs.python.org/3/library/resource.html) module did a better job of measuring memory usage here: 

```python
resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
```

My actual memory issue ended up being me misunderstanding memory usage of `vmap` vs. `lax.map`. (Looks like this will be clarified in #2940!)
